### PR TITLE
Riak Tests for scenarios of basho/riak_kv#679

### PR DESCRIPTION
**Pull request by russelldb**
Note that the fallback test `kv679_dataloss_fb` will fail until that case is addressed with a riak_object format change.

### CODE REVIEW

**Comment by russelldb** [comment]
hmmmm, yes. Can I force push if needed?

**Comment by russelldb** [comment]
I did it anyway, sorry

**Comment by seancribbs** [comment]
Nice, thanks! Ping when you're ready for review.

**Comment by russelldb** [comment]
ping @seancribbs, ready for review

File: tests/kv679_dataloss.erl [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    VCE0 = riakc_obj:vclock(Bod),
+    VC0 = rpc:call(Node, riak_object, decode_vclock, [VCE0]),
+    lager:info("VC ~p~n", [VC0]),
+
+
+    %% delete the local data for Key
+    %% ERM why not just stop the node, then delete the data dir?
+    delete_datadir(hd(PL)),
+
+    timer:sleep(10000),
```
Comment (andrewjstone) [inline tests/kv679_dataloss.erl]: This timeout sticks out as something that may not be helpful for repeatability. Are we sure it's even necessary here?

**Comment by andrewjstone** [comment]
- [x] kv679_dataloss passes with PR, fails without
- [x] kv679_dataloss code review
- [ ] kv679_dataloss_fb passes with PR, fails without
- [ ] kv679_dataloss_fb code review
- [ ] kv679_tombstone passes with PR, fails without
- [ ] kv679_tombstone code review
- [ ] kv679_tombstone2 passes with PR, fails without
- [ ] kv679_tombstone2 code review

File: tests/kv679_dataloss_fb.erl [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    ?assertMatch({ok, _}, Res),
+    {ok, O} = Res,
+
+    %% A nice riak would have somehow managed to make a sibling of the
+    %% last write
+    ?assertEqual([<<"anne">>, <<"joe">>], riakc_obj:get_values(O)),
+
+    pass.
+
+two_fallbacks_one_primary(PL) ->
```
Comment (andrewjstone) [inline tests/kv679_dataloss_fb.erl]: For clarity, I'd change this function name to `primary_and_fallback_counts/1` to indicate it's actual function and not the expected result.

**Comment by andrewjstone** [comment]
Seeing the following failure with `kv679_dataloss_fb`

```
19:45:30.878 [info] Killed 2 primaries
19:45:30.883 [info] Got a preflist with coord and 2 fbs [{{319703483166135013357056057156686910549735243776,'dev3@127.0.0.1'},primary},{{342539446249430371453988632667878832731859189760,'dev6@127.0.0.1'},fallback},{{365375409332725729550921208179070754913983135744,'dev1@127.0.0.1'},fallback}]
19:45:31.202 [info] writing new!
19:45:31.534 [info] writing existing!
19:45:31.536 [info] VC [{<<157,62,8,202,207,232,52,51,0,0,0,1>>,{2,63588501931}}]
19:45:31.536 [info] Coord Node ID <<157,62,8,202>>
19:45:31.536 [info] Clock at 2 fallbacks
19:45:31.536 [info] Killing node 'dev6@127.0.0.1'
19:45:31.536 [info] Killing node 'dev1@127.0.0.1'
19:45:31.537 [info] Running: /Users/ajs/rt/riak_ee/current/dev/dev4/bin/riak start
19:45:34.764 [info] Wait until 'dev4@127.0.0.1' is pingable
19:45:35.174 [info] Wait until 'dev4@127.0.0.1' is pingable
19:45:35.174 [info] Running: /Users/ajs/rt/riak_ee/current/dev/dev5/bin/riak start
19:45:38.389 [info] Wait until 'dev5@127.0.0.1' is pingable
19:45:38.776 [info] Wait until 'dev5@127.0.0.1' is pingable
19:45:38.776 [info] started primaries back up
19:45:48.789 [info] writing existing!
19:45:48.833 [info] writing existing!
19:45:48.835 [info] VC [{<<157,62,8,202,207,232,52,51,0,0,0,1>>,{4,63588501948}}]
19:45:48.836 [info] Coord Node ID <<157,62,8,202>>
19:45:48.836 [info] Killing node 'dev4@127.0.0.1'
19:45:48.836 [info] Killing node 'dev5@127.0.0.1'
19:45:48.836 [info] killed primaries again
19:45:48.838 [info] deleting backend data dir for 319703483166135013357056057156686910549735243776 'dev3@127.0.0.1' on primary
19:45:48.839 [info] Path "/Users/ajs/rt/riak_ee/current/dev/dev3/data/bitcask/319703483166135013357056057156686910549735243776"
19:45:48.839 [info] Stopping riak on 'dev3@127.0.0.1'
19:45:58.840 [info] Running: /Users/ajs/rt/riak_ee/current/dev/dev3/bin/riak stop
19:46:02.498 [info] Wait until 'dev3@127.0.0.1' is not pingable
19:46:02.499 [info] Running: /Users/ajs/rt/riak_ee/current/dev/dev3/bin/riak start
19:46:05.717 [info] Wait until 'dev3@127.0.0.1' is pingable
19:46:06.081 [info] Wait until 'dev3@127.0.0.1' is pingable
19:46:06.081 [info] Running: /Users/ajs/rt/riak_ee/current/dev/dev6/bin/riak start
19:46:09.347 [info] Wait until 'dev6@127.0.0.1' is pingable
19:46:09.722 [info] Wait until 'dev6@127.0.0.1' is pingable
19:46:09.722 [info] Running: /Users/ajs/rt/riak_ee/current/dev/dev1/bin/riak start
19:46:13.026 [info] Wait until 'dev1@127.0.0.1' is pingable
19:46:13.393 [info] Wait until 'dev1@127.0.0.1' is pingable
19:46:13.393 [info] restart fallbacks
19:46:23.529 [info] VC [{<<157,62,8,202,207,232,52,51,0,0,0,1>>,{2,63588501931}}]
19:46:23.529 [info] Coord Node ID <<157,62,8,202>>
19:46:23.530 [info] writing existing!
19:46:23.588 [info] VC [{<<157,62,8,202,207,232,52,51,0,0,0,1>>,{3,63588501983}}]
19:46:23.588 [info] Coord Node ID <<157,62,8,202>>
19:46:23.588 [info] Running: /Users/ajs/rt/riak_ee/current/dev/dev4/bin/riak start
19:46:26.850 [info] Wait until 'dev4@127.0.0.1' is pingable
19:46:27.237 [info] Wait until 'dev4@127.0.0.1' is pingable
19:46:27.237 [info] Running: /Users/ajs/rt/riak_ee/current/dev/dev5/bin/riak start
19:46:30.501 [info] Wait until 'dev5@127.0.0.1' is pingable
19:46:30.878 [info] Wait until 'dev5@127.0.0.1' is pingable
19:46:30.878 [info] restart primaries _again_
19:46:40.888 [info] wait for handoffs
19:46:40.888 [info] Wait until transfers complete 'dev6@127.0.0.1'
19:50:01.381 [info] Wait until transfers complete 'dev1@127.0.0.1'
19:50:01.397 [info] final get
19:50:01.398 [error] Error in process <0.158.0> on node 'riak_test@127.0.0.1' with exit value: {{assertEqual_failed,[{module,kv679_dataloss_fb},{line,179},{expression,"riakc_obj : get_values ( O )"},{expected,[<<4 bytes>>,<<3 bytes>>]},{value,[<<3 bytes>>]}]},[{kv679_dataloss_fb,'-confirm/0-fun-16-',2,[{file,"tests/kv679_dataloss_fb.erl"},{line,179}]},...


19:50:01.398 [warning] kv679_dataloss_fb failed: {{assertEqual_failed,[{module,kv679_dataloss_fb},{line,179},{expression,"riakc_obj : get_values ( O )"},{expected,[<<"anne">>,<<"joe">>]},{value,[<<"joe">>]}]},[{kv679_dataloss_fb,'-confirm/0-fun-16-',2,[{file,"tests/kv679_dataloss_fb.erl"},{line,179}]},{kv679_dataloss_fb,confirm,0,[{file,"tests/kv679_dataloss_fb.erl"},{line,179}]},{riak_test_runner,return_to_exit,3,[{file,"src/riak_test_runner.erl"},{line,159}]}]}
19:50:01.399 [error]
================ kv679_dataloss_fb failure stack trace =====================
{{assertEqual_failed,[{module,kv679_dataloss_fb},
                      {line,179},
                      {expression,"riakc_obj : get_values ( O )"},
                      {expected,[<<"anne">>,<<"joe">>]},
                      {value,[<<"joe">>]}]},
 [{kv679_dataloss_fb,'-confirm/0-fun-16-',2,
                     [{file,"tests/kv679_dataloss_fb.erl"},{line,179}]},
  {kv679_dataloss_fb,confirm,0,
                     [{file,"tests/kv679_dataloss_fb.erl"},{line,179}]},
  {riak_test_runner,return_to_exit,3,
                    [{file,"src/riak_test_runner.erl"},{line,159}]}]}
============================================================================

19:50:01.399 [notice] kv679_dataloss_fb Test Run Complete
19:50:01.401 [info] Stopping cover

Test Results:
kv679_dataloss_fb-bitcask: fail
---------------------------------------------
1 Tests Failed
0 Tests Passed
That's 0.0% for those keeping score
```

File: tests/kv679_tombstone.erl [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+    end.
+
+kill_node(Node) ->
+    rt:stop_and_wait(Node).
+
+wait_for_new_pl(PL, Node) ->
+    rt:wait_until(fun() ->
+                          NewPL = get_preflist(Node),
+                          lager:info("new ~p~n old ~p~nNode ~p~n", [NewPL, PL, Node]),
+                          NewPL /= PL%% andalso contains_fallback(NewPL)
```
Comment (andrewjstone) [inline tests/kv679_tombstone.erl]: I agree that you don't need the `andalso contains_fallback(NewPL)`. In this case the comment should be removed and so should `contains_fallback/1`.

File: tests/kv679_tombstone2.erl [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+    delete_key(CoordClient),
+
+    lager:info("deleted key"),
+
+    %% kill the patsy, must happen before the reap
+    %%
+    %% @TODO make deterministic (would be nice without all these
+    %% timers)
+    timer:sleep(500),
```
Comment (andrewjstone) [inline tests/kv679_tombstone2.erl]: I understand the sleep after killing the Patsy, but why this one?

File: tests/kv679_tombstone2.erl [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+-compile([export_all]).
+-export([confirm/0]).
+
+-include_lib("eunit/include/eunit.hrl").
+
+-define(BUCKET, <<"kv679">>).
+-define(KEY, <<"test">>).
+
+confirm() ->
+    Config = [{riak_kv, [{delete_mode, 10000}]}, %% 20 seconds to reap.
```
Comment (andrewjstone) [inline tests/kv679_tombstone2.erl]: Think the comment should say `10 seconds to reap`

File: tests/kv679_tombstone2.erl [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    %% timers)
+    timer:sleep(500),
+
+    rt:brutal_kill(Patsy),
+
+    lager:info("killed the patsy"),
+
+    %% A time to reap
+    %% %% wait for the up  nodes to reap
+    timer:sleep(15000),
```
Comment (andrewjstone) [inline tests/kv679_tombstone2.erl]: Why not just use `kv679_tombstone:read_it_and_reap/1` or something like that to make this deterministic?

File: tests/kv679_tombstone2.erl [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+    lager:info("tombstone (should be) reaped"),
+
+    %% %% write the key again, this will start a new clock, a clock
+    %% that is in the past of that un-reaped primary tombstone. We use the
+    %% same node to get the same clock.
+    kv679_tombstone:write_key(CoordClient, [<<"jon">>],
+                              [{pw, 2},
+                               {sloppy_quorum, false},
+                               {n_val, 2}]),
```
Comment (andrewjstone) [inline tests/kv679_tombstone2.erl]: Why bother turning off sloppy quorums and only writing to primaries here? Also, if you write this request with an n_val of 2 here, will  this key still be replicated to all 3 primaries on read repair?

**Comment by russelldb** [comment]
WRT the failing `kv679_dataloss_fb` that scenario needs the `dirty_clock` stuff we discussed that is `@TODO`'d in `riak_kv_vnode`. Should I remove that test from this PR, and re-submit it when we finally fix that case?

File: tests/kv679_dataloss.erl (hunk 1 above)
Comment (russelldb) [inline tests/kv679_dataloss.erl]: It's not, vestige of early iteration. Good catch.

File: tests/kv679_tombstone2.erl (hunk 4 above)
Comment (russelldb) [inline tests/kv679_tombstone2.erl]: No idea, test passes with out it, again, probably a vestige from my blind groping when trying to make these tests.

File: tests/kv679_tombstone2.erl (hunk 6 above)
Comment (russelldb) [inline tests/kv679_tombstone2.erl]: There's a reason…I wish I could remember. Reading it causes read repair, and `read_it_and_reap` reads the key until it's reaped, here we just wait for the magic to happen.

File: tests/kv679_tombstone2.erl (hunk 7 above)
Comment (russelldb) [inline tests/kv679_tombstone2.erl]: There's a reason, and I really wish I'd made a note of it in the comments. Probably to do with fallbacks, but let me try without all this and see.

Comment (russelldb) [inline tests/kv679_tombstone2.erl]: I tried without this config, there is no reason. Again, some left over cargo-cultage from early attempts to reproduce the theoretical issue in practice.

File: tests/kv679_tombstone2.erl (hunk 6 above)
Comment (russelldb) [inline tests/kv679_tombstone2.erl]: Hmm, again, works fine using `read_it_and_reap` more vestigial remains of groping. This review process works, eh?

Comment (russelldb) [inline tests/kv679_tombstone2.erl]: Ok, hmm, ok. Test passes against 2.0.4 with `read_it_and_reap`. This is why I used the timer. Not sure of the reason yet: is it the tombstone getting read_repaired onto a fallback? Something else? Do I _have_ to know for the review to pass?

**Comment by russelldb** [comment]
Hopefully that does it. Let me know if about removing `kv679_dataloss_fb` until we fix that.

**Comment by andrewjstone** [comment]
@russelldb Great job here! I would remove `kv679_dataloss_fb` if it's not going to pass. Just put it in another PR for when we add the dirty clock support to the code.

+1 after that.

**Comment by andrewjstone** [comment]
:+1:

**Pull request merged**